internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,674 @@
+package graph
+
+import (
+	"context"
+	"slices"
+
+	"github.com/emirpasic/gods/sets/hashset"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/internal/checkutil"
+	"github.com/openfga/openfga/internal/concurrency"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+const IteratorMinBatchThreshold = 1000
+
+type fastPathSetHandler func(ctx context.Context, iterQueue []*iteratorProducer, iterChan chan *iteratorMsg)
+
+type iteratorMsg struct {
+	iter storage.TupleKeyIterator
+	err  error
+}
+
+type iteratorProducer struct {
+	idx      int
+	iter     storage.TupleKeyIterator
+	done     bool
+	producer chan *iteratorMsg
+}
+
+// NOTE: caller should consider running this in a goroutine to not block.
+func cleanupIteratorProducers(iterProducers []*iteratorProducer) {
+	for _, iterProducer := range iterProducers {
+		if iterProducer.iter != nil {
+			iterProducer.iter.Stop()
+		}
+		for msg := range iterProducer.producer {
+			if msg.iter != nil {
+				msg.iter.Stop()
+			}
+		}
+	}
+}
+
+// pollIteratorProducers will return a list of the remaining active producers.
+// To be considered active your producer channel must still be open.
+func pollIteratorProducers(ctx context.Context, iterProducers []*iteratorProducer) ([]*iteratorProducer, error) {
+	for _, producer := range iterProducers {
+		if producer.iter != nil || producer.done {
+			// no need to poll further
+			continue
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case i, ok := <-producer.producer:
+			if !ok {
+				producer.done = true
+				break
+			}
+			if i.err != nil {
+				return nil, i.err
+			}
```

Comment:
could you add a godoc here explaining why we return immediately? what will happen to the other possibly active iterators? they will not be returned